packages/babel-parser/src/parser/statement.ts

```diff
@@ -1719,7 +1719,16 @@ export default abstract class StatementParser extends ExpressionParser {
         }
 
         if (this.match(tt.at)) {
-          decorators.push(this.parseDecorator());
+          // All parts of a ClassDeclaration or a ClassExpression are strict mode code,
+          // except for the DecoratorList at the top level of a ClassDeclaration or ClassExpression which is contained in non - strict mode code.
+          if (!oldStrict && this.hasPlugin("decorators")) {
+            this.state.strict = false;
+            decorators.push(this.parseDecorator());
+            this.state.strict = true;
+          } else {
+            decorators.push(this.parseDecorator());
+          }
+
```

Comment:
According to my understanding, the attribute decorator in the class is an exception, does not follow the mandatory strict mode, but will inherit the upper level of the class.

https://github.com/babel/babel/pull/15243#issuecomment-1334692759

Reply:
@liuxingbaoyu 
> the attribute decorator in the class is an exception, does not follow the mandatory strict mode, but will inherit the upper level of the class

I don't think so. Here is an example of "DecoratorList at the top level of a ClassDeclaration or ClassExpression"

```js
// sourceType: "script"
function f() {
  @yield() class C {}
}
```
Although `class C {}` should be parsed in strict mode, `@yield ()` inherits the upper non-strict mode from the function body. Babel parser already worked: https://astexplorer.net/#/gist/562b73cfaf64bdaa2bff9d5958b9cec2/708bb6291e630c8482f5448d9a6887b95306075f

But this is invalid, because all parts of ClassDeclaration are strict mode
```js
// sourceType: "script"
function f() {
  class C { @yield() f }
}
```

Reply:
Seems plausible, I didn't find any more useful information in the proposal repo.
Changes from
https://github.com/tc39/test262/commit/e7895c53d295760d7c2b89743f8311fa60480392 , no discussion about it was found.

Also I have another question, is await legal here? This is also from test262.
https://babeljs.io/repl#?browsers=&build=&builtIns=false&corejs=3.21&spec=false&loose=false&code_lz=MYGwhgzhAEDC0G8BQ1oAEAUYDuYCWALgJTQC2ApgQBYD2AJhiQgL4rpa6EkQFgF7AylWgyatUmHPmLQAZnnIg6AbjaTOMnnwFyFS1cyA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=script&lineWrap=true&presets=stage-2&prettier=false&targets=&version=7.20.6&externalPlugins=&assumptions=%7B%7D

Reply:
Good question. Per

https://tc39.es/ecma262/#prod-IdentifierReference

`await` is allowed in `[~Await]` productions. Since the `DecoratorList` inherits `Await` production, `await` can be used in decorator as long as it is not derived from `[+Await]` productions. For example,

```js
function f() {
  @await class C { @await p }
}
```
is valid, while
```js
// sourceType: "module"
@await class C { @await p }
```
is not.

Reply:
```
class C {
  @(await) field;
}

class C2 {
  @(yield) field;
}
```
The specification seems to refer to `await` and `yield` as similar things, but such code `yield` will fail. I suspect this is being affected by `topLevelAwait`, is this expected?

Reply:
Well `await` is not allowed as an identifier reference in module sourceType, so

```js
// sourceType: "module"
function f() { class C { @(await) f } }
```

is invalid, even if it is within an non-async context. See also https://tc39.es/ecma262/#sec-identifiers-static-semantics-early-errors.

BTW I have open https://github.com/tc39/test262/pull/3744 to remove those cases.

Reply:
Is there a difference between `await` and `yield` in strict mode of `script`?
Now the `await` decorator in the class can be successfully parsed by babel, but `yield` cannot.

Thank you for opening that PR, I'm not familiar with test262!

Reply:
![image](https://user-images.githubusercontent.com/30521560/205388156-a332b833-a18a-4692-b2b6-58cf9cf3baa5.png)
A simpler example.

Reply:
> Is there a difference between await and yield in strict mode of script?

`await` is disallowed in `[Await]` while `yield` is disallowed in `[Yield]`. You can refer to https://tc39.es/ecma262/#sec-identifiers-static-semantics-early-errors.

Reply:
![image](https://user-images.githubusercontent.com/30521560/205391003-ef1f5e5e-9ca0-4812-a222-cbf582496bca.png)
I get it, they differ here. Thanks!